### DISCUSSION TITLE
[Android] Self-shadowing not working

### DISCUSSION THREAD

**Discussion by wes-nz** [google/filament#4590] [Q&A] [answered]
Hi,

I am struggling to get self-shadowing working on Android. I am using a copy of `ModelViewer.kt`.

My Light 
```Kotlin
        LightManager.Builder(LightManager.Type.FOCUSED_SPOT)
            .color(255f, 255f, 255f)
            .falloff(80f)
            .intensity(150_000.0f)
            .position(0f, 5f, -4f)
            .castShadows(true)
            .build(engine, lightOne)
```

When the scene is populated (same code as `populateScene`) except
```Kotlin
rcm.setScreenSpaceContactShadows(ri, true)
rcm.setCastShadows(ri, true)
rcm.setReceiveShadows(ri, true)
```

As you can see the nose or glasses don't cast any shadow in Filament.
![https://user-images.githubusercontent.com/2945380/132118062-85de369f-b770-4fc9-95a9-107f9343e5cd.mp4
](https://user-images.githubusercontent.com/2945380/132118062-85de369f-b770-4fc9-95a9-107f9343e5cd.mp4)

SceneKit
![https://user-images.githubusercontent.com/2945380/132118325-32637347-58a8-478d-9dd6-845653bed67c.mp4](https://user-images.githubusercontent.com/2945380/132118325-32637347-58a8-478d-9dd6-845653bed67c.mp4)

Am I missing something obvious?

Thank you for all your work on the engine 🙏

**Comment by romainguy** [comment, marked as answer]
Note that the color of a light is not specified as values between 0 and 255 but 0 and 1. You are making your light 255 times brighter. If that doesn't fix it you'll have to share more of your code. glTF files are setup to cast and receive shadows automatically.

**Reply by wes-nz** [reply]
Good catch, unfortunately no luck. 

I made a dummy project with the bare minimum. The `ModelViewer` is a copy paste from the library except the lines mentioned above. [The project is available for download here](https://www.dropbox.com/s/vn66r0835t61vz6/selfshadow.zip?dl=0). In that example project I had to bump the light intensity to `3_000_000` which seem very odd... 

Thank you for your time.

**Reply by romainguy** [reply]
A spot shadow's intensity is defined in lumens, and the exposure settings you use in `ModelViewer` are for bright sunny outdoors (ƒ/16 1/125s ISO 100). That's why you need an extremely bright spot light. You should change your exposure instead.

**Reply by wes-nz** [reply]
Ah of course, you're right, after testing that explains the lighting. I am still not sure why the shadows aren't visible.

**Reply by romainguy** [reply]
I'm looking into it right now. Spot lights cast shadows just fine in our desktop test apps, I'm not sure what's going on here. @bejado, in the test app above, shadows work fine on a directional light but not on spot lights. Any idea?

Note: enabling screen space contact shadows works.

**Reply by bejado** [reply]
I'll take a closer look this week.

**Reply by bejado** [reply]
Strange, I'm not able to get any shadows with spot lights, let alone self-shadows. Maybe something broke recently.

**Reply by romainguy** [reply]
@pixelflinger Did you make changes recently that could have affected shadows?

**Reply by bejado** [reply]
Ah I figured it out. The `spotLightCone` needs to be explicitly set. Apparently we don't have good defaults for it:

```
        LightManager.Builder(LightManager.Type.FOCUSED_SPOT)
                .color(255f, 255f, 255f)
                .falloff(10f)
                .intensity(3500.0f)                                   // back off the intensity
                .position(0f, 5f, -4f)
                .spotLightCone(3.14159f / 32f, 3.14159f / 32f + .1f)  // <--
                .castShadows(true)
                .build(engine, light)
```

**Reply by romainguy** [reply]
That's odd it's exactly what I did and I didn't get shadows either. Maybe I didn't test correctly. Would you mind changing the default btw?

**Reply by wes-nz** [reply]
Thanks @bejado, have you changed anything else? I am struggling to get the desired result with your changes inside the test project.

**Reply by bejado** [reply]
No, that was the only change I needed to make to get shadows working. You may also need to play around with the IBL / sun intensity to be able to see them.

**Reply by wes-nz** [reply]
Ehm I tried that too... By any chance could you zip the project back? Thanks.

**Reply by bejado** [reply]
Here's a branch that contains the change: https://github.com/google/filament/tree/bjd/spot-light-shadow-demo
![Screenshot_20210914-114433](https://user-images.githubusercontent.com/5298046/133316612-d20bcebb-7b9c-40ac-9765-dca176391db7.png)

**Reply by wes-nz** [reply]
I see, I had to change the values for the cones in order to see the shadows: `.spotLightCone(3.14159f / 36f, 3.14159f / 3f)`. 
Thanks everyone for the help, much appreciated.
